### PR TITLE
CancellationTokenSource in Cahe TagHelpers

### LINKED ISSUE #5649: CancellationTokenSource is not Disposed in cache tag helpers
Inside https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs#L80 (ProcessAsync method) there is a CancellationTokenSource instantiation without proper Dispose call.

Also, https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.TagHelpers/DistributedCacheTagHelper.cs#L72 creates a new CancellationTokenSource which is not used anywhere.

I believe these places should be fixed.

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #5649

### CODE REVIEW

**Comment by sebastienros** [comment]
Bonus points to myself for creating these commits withing GitHub 🍰

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs [hunk 1]
```diff
@@ -122,6 +122,7 @@ public CacheTagHelper(IMemoryCache memoryCache, HtmlEncoder htmlEncoder) : base(
                             // If an exception occurs, ensure the other awaiters
                             // render the output by themselves.
                             tcs.SetResult(null);
+                            tokenSource.Dispose();
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs]: Does anything happen to `tokenSource.Token` or the `CancellationChangeToken` containing it when the source is disposed?

**Comment by Eilon** [comment]
Ping @sebastienros

**Comment by sebastienros** [comment]
I spent some time looking at the implementations of `CancellationToken` and `CancellationTokenSource`. H?ow we use the token in the `MemoryCache` will never require the source from checking if it's disposed. The `CancellationChangeToken` that links to the token will only call into `Token.IsCancellationRequested` which itself checks its source (disposed) but only checks a flag. So even this will just work.

TL;DR: The token links to its disposed source, but no current code path will fail from it.

**Comment by sebastienros** [comment]
On the other side, based on our usage of the `CancellationTokenSource` there will be no `Timer` or callbacks registers, which means there is really nothing to dispose. There is also potentially a `ManualResetEvent` but I don't think it's initialized. So it should be safe in terms of memory management to not call `Dispose`.

My gut feeling would be to actually not call `Dispose` until it's proven that there is a memory leak from it, with our specific usage.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs [hunk 2]
```diff
@@ -120,6 +120,7 @@ public CacheTagHelper(IMemoryCache memoryCache, HtmlEncoder htmlEncoder) : base(
                             // If an exception occurs, ensure the other awaiters
                             // render the output by themselves.
                             tcs.SetResult(null);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs]: Complete with content in the success case.

**Comment by sebastienros** [comment]
After discussing with @Tratcher it's necessary to call `Dispose` as the memory cache will register a callback. But this callback won't be invoked as the Token is not accessed later on.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs [hunk 3]
```diff
@@ -108,18 +108,24 @@ public CacheTagHelper(IMemoryCache memoryCache, HtmlEncoder htmlEncoder) : base(
 
                                 content = await result;
                             }
+                            
+                            tcs.SetResult(content);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs]: Looks like setting this might be short-circuiting an extra (unnecessary loop). Consider this:

1. Caller A gets a cache miss and enters the critical sections sets a TCS.Task as the Cache Entry
2. Caller B does a TryGetValue and gets the TCS.Task
3. Caller A replaces the CacheEntry.Value with the task returned from the TagHelper.
4. Caller A does TCS.SetResult(null);
5. Caller A's output is valid html. Caller B's output is null. It has to redo the `TryGetValue` at step 2.

It's pretty unrelated to the actual change, but might help us remove the `while` loop.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/CacheTagHelperTest.cs [hunk 4]
```diff
@@ -743,15 +742,63 @@ Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEncoder en
             Assert.Equal(childContent, tagHelperOutput4.Content.GetContent());
         }
 
+        [Fact]
+        public async Task ProcessAsync_WorksForNestedCacheTagHelpers()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/CacheTagHelperTest.cs]: Verifies we aren't attempting to register a call back on a disposed token.

**Comment by pranavkm** [comment]
@Tratcher - could you have a look?

**Comment by rynowak** [comment]
:shipit:
